autogluon/utils/tabular/ml/models/fastainn/tabular_nn_fastai.py

```diff
@@ -0,0 +1,334 @@
+            self.col_after_transformer = [col for col in X.columns if col in cols_to_use]
+        X = pd.DataFrame(data=X, columns=self.col_after_transformer)
+        X = super().preprocess(X)
+        return X
+
+    def preprocess_train(self, X_train, Y_train, X_test, Y_test, **kwargs):
+        from fastai.data_block import FloatList
+        from fastai.tabular import TabularList
+        from fastai.tabular import FillMissing, Categorify, Normalize
+
+        self.cat_columns = X_train.select_dtypes(['category', 'object']).columns.values.tolist()
+        self.cont_columns = X_train.select_dtypes(['float', 'int', 'datetime']).columns.values.tolist()
+        if self.problem_type == REGRESSION and self.y_scaler is not None:
+            Y_train_norm = pd.Series(self.y_scaler.fit_transform(Y_train.values.reshape(-1, 1)).reshape(-1))
+            Y_test_norm = pd.Series(self.y_scaler.transform(Y_test.values.reshape(-1, 1)).reshape(-1)) if Y_test is not None else None
+            logger.log(0, f'Training with scaled targets: {self.y_scaler} - !!! NN training metric will be different from the final results !!!')
+        else:
+            Y_train_norm = Y_train
+            Y_test_norm = Y_test
+        try:
+            X_train_stats = X_train.describe(include='all').T.reset_index()
+            cat_cols_to_drop = X_train_stats[(X_train_stats['unique'] > self.max_unique_categorical_values) | (X_train_stats['unique'].isna())]['index'].values
+        except:
+            cat_cols_to_drop = []
+        cat_cols_to_keep = [col for col in X_train.columns.values if (col not in cat_cols_to_drop)]
+        cat_cols_to_use = [col for col in self.cat_columns if col in cat_cols_to_keep]
+        logger.log(15, f'Using {len(cat_cols_to_use)}/{len(self.cat_columns)} categorical features')
+        self.cat_columns = cat_cols_to_use
+        self.cat_columns = [feature for feature in self.cat_columns if feature in list(X_train.columns)]
+        self.cont_columns = [feature for feature in self.cont_columns if feature in list(X_train.columns)]
+        logger.log(15, f'Using {len(self.cont_columns)} cont features')
+        X_train = self.fold_preprocess(X_train, fit=True)
+        if X_test is not None:
+            X_test = self.fold_preprocess(X_test)
+        df_train, train_idx, val_idx = self._generate_datasets(X_train, Y_train_norm, X_test, Y_test_norm)
+        label_class = FloatList if self.problem_type == REGRESSION else None
+        procs = [FillMissing, Categorify, Normalize]
+        data = (TabularList.from_df(df_train, path=self.path, cat_names=self.cat_columns, cont_names=self.cont_columns, procs=procs)
+                .split_by_idxs(train_idx, val_idx)
+                .label_from_df(cols=LABEL, label_cls=label_class)
+                .databunch(bs=self.params['bs'] if len(X_train) > self.params['bs'] else 32))
+        return data
+
+    def _fit(self, X_train, y_train, X_val=None, y_val=None, time_limit=None, **kwargs):
```

Comment:
`kwargs` may contain user-specified desire to use GPU.  Can we support this for this model?

Here's an example for the Gluon network:  https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py#L196

Reply:
fastai automatically detects and uses GPU if present. I added TODO to handle case when CUDA is available, but `num_gpus` is set to 0.

Reply:
In future we should rework the gpu flag to be part of AG_args_fit, it shouldn't be in kwargs.